jupyterhub/app.py

```diff
@@ -2009,22 +2047,35 @@ async def init_roles(self):
                                 )
                         Class = orm.get_class(bearer)
                         orm_obj = Class.find(db, bname)
-                        roles.grant_role(
-                            db, entity=orm_obj, rolename=predef_role['name']
-                        )
-        # make sure that on no admin situation, all roles are reset
-        admin_role = orm.Role.find(db, name='admin')
-        if not admin_role.users:
+                        orm_role_bearers.append(orm_obj)
+                        # Ensure all with admin role have admin flag
+                        if predef_role['name'] == 'admin':
+                            orm_obj.admin = True
+                setattr(predef_role_obj, bearer, orm_role_bearers)
+                db.commit()
+        allowed_users = db.query(orm.User).filter(
+            orm.User.name.in_(self.authenticator.allowed_users)
+        )
+        for user in allowed_users:
+            roles.grant_role(db, user, 'user')
+        admin_role = orm.Role.find(db, 'admin')
+        for bearer in admin_role_objects:
+            Class = orm.get_class(bearer)
+            for admin_obj in db.query(Class).filter_by(admin=True):
+                if has_admin_role_spec[bearer]:
+                    admin_obj.admin = admin_role in admin_obj.roles
+                else:
+                    roles.grant_role(db, admin_obj, 'admin')
+        db.commit()
+        # make sure that on hub upgrade, all roles are reset
+        if not getattr(self, '_rbac_upgrade', False):
```

Comment:
This is backwards, right?

```suggestion
        if getattr(self, '_rbac_upgrade', False):
```

If this is backwards, it also means it's untested, which should probably be fixed.

Reply:
Actually, I just ran into that because some tests were failing. But I will add an explicit test for it, thanks